crda/remotematcher.go

```diff
@@ -0,0 +1,258 @@
+}
+
+// Option controls the configuration of a Matcher.
+type Option func(*Matcher) error
+
+// NewMatcher returns a configured Matcher or reports an error.
+func NewMatcher(opt ...Option) (*Matcher, error) {
+	m := Matcher{}
+	for _, f := range opt {
+		if err := f(&m); err != nil {
+			return nil, err
+		}
+	}
+
+	if m.url == nil {
+		var err error
+		m.url, err = url.Parse(defaultURL)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if m.client == nil {
+		m.client = http.DefaultClient
+	}
+	// defaults to a sane concurrency limit.
+	if m.requestConcurrency < 1 {
+		m.requestConcurrency = defaultRequestConcurrency
+	}
+
+	return &m, nil
+}
+
+// WithClient sets the http.Client that the matcher should use for requests.
+//
+// If not passed to NewMatcher, http.DefaultClient will be used.
+func WithClient(c *http.Client) Option {
+	return func(m *Matcher) error {
+		m.client = c
+		return nil
+	}
+}
+
+// WithHost sets the server host name that the matcher should use for requests.
+//
+// If not passed to NewMatcher, defaultHost will be used.
+func WithURL(uri string) Option {
+	u, err := url.Parse(uri)
+	return func(m *Matcher) error {
+		if err != nil {
+			return err
+		}
+		m.url = u
+		return nil
+	}
+}
+
+// WithRequestConcurrency sets the concurrency limit for the network calls.
+//
+// If not passed to NewMatcher, a defaultRequestConcurrency will be used.
+func WithRequestConcurrency(requestConcurrency int) Option {
+	return func(m *Matcher) error {
+		m.requestConcurrency = requestConcurrency
+		return nil
+	}
+}
+
+// Name implements driver.Matcher.
+func (*Matcher) Name() string { return "crda" }
+
+// Filter implements driver.Matcher.
+func (*Matcher) Filter(record *claircore.IndexRecord) bool {
+	return record.Package.NormalizedVersion.Kind == "pep440"
+}
+
+// Query implements driver.Matcher.
+func (*Matcher) Query() []driver.MatchConstraint {
+	panic("unreachable")
+}
+
+// Vulnerable implements driver.Matcher.
+func (*Matcher) Vulnerable(ctx context.Context, record *claircore.IndexRecord, vuln *claircore.Vulnerability) (bool, error) {
+	// RemoteMatcher can match Package and Vulnerability.
+	panic("unreachable")
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "crda/remotematcher/QueryRemoteMatcher").
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Debug().
+		Int("records", len(records)).
+		Msg("packages")
+
+	ctrlC, errorC := m.fetchVulnerabilities(ctx, records)
+	results := make(map[string][]*claircore.Vulnerability)
+	for r := range ctrlC {
+		for _, vuln := range r {
+			results[vuln.Package.ID] = append(results[vuln.Package.ID], vuln)
+			log.Debug().
+				Str("package", vuln.Package.Name).
+				Str("version", vuln.Package.Version).
+				Str("id", vuln.ID).
+				Msg("vulns")
+		}
+	}
+	select {
```

Comment:
I'm not sure this select statement is going to work as intended. Checking "ok" only tells you if the error channel has closed, not whether an error has returned or not. 

Since your code guarantees errorC will be closed, I think you simply want to block on the error channel, then check the error.

```
err := <-errorC // guaranteed to have an err or be closed
if err != nil {
   	log.Error().Err(err).Msg("access to component analyses failed")
}
```

Also to guarantee  that ctrlC is closed before errorC you can swap your defers on lines 185,186
```
defer close(errorC)
defer close(ctrlC)
```

these defers stack, so ctrlC will be closed, and then errorC. its not explicitly necessary but it aligns with the order of operations in QuayRemoteMatcher.

Reply:
Ack!

Reply:
Fixed it.